keps/sig-network/3766-referencegrant/README.md

```diff
@@ -0,0 +1,492 @@
+Gateway API conformance tests can serve as a model for this.
+
+#### Cross-Namespace References may Weaken Namespace Boundaries
+Although we believe that the handshake required for cross-namespace references
+with ReferenceGrant ensures these references will be safe, it does potentially
+weaken existing namespace boundaries. We believe ReferenceGrant will have a net
+benefit on the ecosystem as it will allow workloads, secrets, and configuration
+to be deployed in separate namespaces that more clearly match up with desired
+authorization.
+
+## Design Details
+
+### General Notes
+
+#### `ReferenceGrant` is half of a handshake
+
+When thinking about ReferenceGrant, it is important to remember that it does not
+do anything by itself. It *Grants* the *possibility* of making a *Reference*
+across namespaces. It's intended that _another object_ (that is, the From object)
+complete the handshake by creating a reference to the referent object (the To
+object).
+
+#### `Kind` vs `Resource`
+
+When creating a metaresource (that is, a resource that targets other resources)
+like ReferenceGrant, it's important to consider if the metaresource uses the more
+common `Kind` or the more correct `Resource`.
+
+When designing the Gateway API in general, we considered this quite a bit (and
+@robscott even ended up making https://github.com/kubernetes/community/pull/5973
+to clarify the API conventions), but in the end decided to use Kind, to improve
+the user experience. That is, it's easier users to take the value from the `kind`
+field at the top of the YAML they are already using, and put it straight into
+these fields, rather than needing to do a kind-resource lookup for every user's
+interaction with the API.
+
+Secondly, as in the Ingress API, the target types must be clearly defined by
+each implementation, so there should be no opportunity for ambiguity.
+
+Lastly, this is only half of a handshake, so the From referent must also both
+support referencing the To referent, and must reference the one in the To field.
+
+With all of that in mind, we felt that the benefits in terms of specificity
+outweighed the costs in terms of user pain to use the more correct `Resource`.
+
+### API Spec
+
+```golang
+// ReferenceGrant identifies kinds of resources in other namespaces that are
+// trusted to reference the specified kinds of resources in the same namespace
+// as the policy.
+//
+// Each ReferenceGrant can be used to represent a unique trust relationship.
+// Additional ReferenceGrants can be used to add to the set of trusted
+// sources of inbound references for the namespace they are defined within.
+//
+// ReferenceGrant is a form of runtime verification allowing users to assert
+// which cross-namespace object references are permitted. Implementations that
+// support ReferenceGrant MUST NOT permit cross-namespace references which have
+// no grant, and MUST respond to the removal of a grant by revoking the access
+// that the grant allowed.
```

Comment:
Can this clarify whether that means existing objects that made use of the cross-namespace relationship should be deactivated/torn down, or that they should coast on the last-granted content? For example, if load balancer A in namespace A uses a TLS secret from namespace B authorized via a reference grant, should removing that grant deactivate the load balancer or prevent it from getting updated content from the secret in namespace B?

Reply:
It seems like this will probably be contextual - I think a TLS cert should be revoked ASAP, but the HTTP route itself would not be torn down.

Reply:
In the existing Gateway API implementation, it is contextual - certs being revoked means the Listener becomes invalid and so is removed, but a cross-namespace backend's ReferenceGrant being revoked means _that backend_ is invalid, which only definitely means _that backend_ is removed. But if it's the only backend, then the removal can cascade outwards.

To answer @liggitt's question explicitly: currently in Gateway API, if user A uses a Gateway in namespace A to reference a Secret in namespace B, and the ReferenceGrant is removed, then the Listener is no longer valid, and must be removed. (Having invalid TLS config is a fatal config error).

I'll add something to say both that:
- the removal must be done as soon as possible (eventual consistency permitting)
- the results of the removal _may_ depend on the exact use the reference is being put to, but _should_ err on the side of removing access.

We talked a lot about where to err here for Gateway API, on keeping things up, or having things be more "secure" (for some value of secure), and came down on the side of more secure. As commented elsewhere here, ReferenceGrant is an access control object, removing it is likely to be a destructive operation. (This is another reason why @robscott and I are both in favor of _some_ form of status communication, whether or not it's a `status` subresouce).

Reply:
I think this is well covered with the latest updates to the KEP, let me know if I missed anything.